### PR TITLE
test: generate core dumps on crashes in debug clusters

### PR DESCRIPTION
**Pull request by avikivity**
The cluster manager library doesn't set the asan/ubsan options to abort on error and create core dumps; this makes debugging much harder.

Fix by preparing the environment correctly.

Fixes scylladb/scylladb#17510

### CODE REVIEW

**Review by nyh — CHANGES_REQUESTED** [review]
This is the wrong fix:

test.py already makes the effort to set UBSAN_OPTIONS / ASAN_OPTIONS environment variables properly (see test.py), and that should affected all of the test suites which use test.py.

And it would have worked, except for the incorrect decision to do `env.clear()` in test/pylib/scylla_server.py! The comment says the intention of this env.clear was to remove SCYLLA_HOME - but it removes everything, including the UBSAN settings!

Recently I had the same discussion with @kostja and I convinced him in another place that we should not clear the entire environment and should only clear specific SCYLLA_* environment variables, and the result was commit 246da8884a519ba886202dcce79de15440c7fc60. But scylla_server.py needs to do the same thing as well.

**Comment by nyh** [comment]
> 
>     * [cql-pytest.test_native_functions.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6994/testReport/junit/%28root%29/non-boost%20tests/cql_pytest_test_native_functions_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_native_functions.debug.1)

This test, test_mintimeuuid_extreme_from_totimestamp, tests some ridiculous date from the age of the dinosaurs used on totimestamp(), and expect to see an error message (which is fine) but not crash (as it did in #17035). As we can see now, the test no longer crashes but it does invoke some UBSAN error and causes a crash because of that. So we'll need to
1. As a quick workaround, "@skip" this test.
2. The real solution will be to fix the code to not get UBSAN errors during this test.

**Comment by nyh** [comment]
Opened https://github.com/scylladb/scylladb/issues/17516 for fixing the problem that UBSAN starts detecting once you enable it.

**Comment by nyh** [comment]
> Opened #17516 for fixing the problem that UBSAN starts detecting once you enable it.

Sent a PR to fix it: https://github.com/scylladb/scylladb/pull/17527

**Comment by avikivity** [comment]
v2: rebased

**Comment by avikivity** [comment]
@nyh please re-review

**Comment by avikivity** [comment]
(see @kostja's comment)

**Comment by nyh** [comment]
> @nyh I don't understand your comment and don't see your PR. 

Yes, my comment was a bit confusing. The PR I linked (https://github.com/scylladb/scylladb/pull/17527) fixes a bug which the ASAN detector detected - once Avi enabled it.
This is old news, and it's not relevant to the current state of Avi's patch.

> test.py prepared ASAN options for run_test, that is, when we run scylla as a test subprocess. Avi's fix patches the case when we run scylla server in client-server environment. Avi's fix is correct and should be merged. Your link leads to some different bug.

I would have preferred a different fix (as I explained above) but I'm not strongly against this one, so I'll merge.

Next time if you positively review a patch (you said "Avi's fix is correct and should be merged"), please use the review buttons on github to give an "approve". It's more concrete than writing positive comments.

**Comment by nyh** [comment]
Actually, I see that no CI ever passed on this so I'll wait for a passing CI before merging this :-(

**Comment by avikivity** [comment]
Maybe we should remove this env.clear().

**Comment by kostja** [comment]
* Avi Kivity ***@***.***> [24/04/01 18:44]:
> Maybe we should remove this env.clear().

test.py tries to sand-box the tests to reduce
flakiness. 

If you remove env.clear(), you may gain some extra features:
e.g. some tests may gain from accepting parameters from the
developer's env.

But some parameters of the developer env may negatively influence
some tests, and it would be hard to debug.

If you want a parameter, you can make an exception for it in
test.py code and mention it in the documentation.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by nyh** [comment]
> Maybe we should remove this env.clear().

This is exactly what I suggested in 
https://github.com/scylladb/scylladb/pull/17511#pullrequestreview-1899763630
as the better fix. I also noted that Kostja already did this in other places - clearing only specific known-problematic variables and leaving the rest to be inherited.
But I committed your patch because Kostja says it's the correct approach in his opinion, and he's the main test.py developer.

**Comment by nyh** [comment]
> * Avi Kivity ***@***.***> [24/04/01 18:44]:
> > Maybe we should remove this env.clear().
> 
> test.py tries to sand-box the tests to reduce
> flakiness. 
> 
> If you remove env.clear(), you may gain some extra features:
> e.g. some tests may gain from accepting parameters from the
> developer's env.
> 
> But some parameters of the developer env may negatively influence
> some tests, and it would be hard to debug.
> 
> If you want a parameter, you can make an exception for it in
> test.py code and mention it in the documentation.
> 
> -- 
> Konstantin Osipov, Moscow, Russia
> 

Actually, test.py can clear the environment it inherits from the user (it shouldn't in my opinion, but that's another story), but the point here is that in environment variables that *it* sets (in this case Asan stuff) it shouldn't clear them again when starting a child process.

**Comment by avikivity** [comment]
Environment variables are a common way to pass out-of-band parameters. Yes, clearing the environment reduces flakiness, but also reduces a user's ability to configure things. Which can result in things not working at all.

**Comment by nyh** [comment]
@avikivity please note that at this point the discussion is theoretical - I already merged your patch and the bug will be fixed. In my opinion you fixed it in a suboptimal way, duplicating the setting of UBSAN_OPTIONS in two places in the code where one would have sufficed, but it's a correct fix nonetheless.

**Pull request closed without merging**
